src/main/java/net/snowflake/client/jdbc/SnowflakeFileTransferAgent.java

```diff
@@ -1268,6 +1287,83 @@ private static JsonNode parseCommandInGS(SFStatement statement, String command)
     return result;
   }
 
+  /**
+   * This is API function to parse the File Transfer Metadatas from a supplied PUT call response.
+   *
+   * <p>NOTE: It only supports PUT on S3/AZURE/GCS
+   *
+   * @param jsonNode JSON doc returned by GS from PUT call
+   * @return The file transfer metadatas for to-be-transferred files.
+   * @throws SnowflakeSQLException if any error occurs
+   */
+  public static List<SnowflakeFileTransferMetadata> getFileTransferMetadatas(JsonNode jsonNode)
+      throws SnowflakeSQLException {
+    CommandType commandType = CommandType.UPLOAD;
+    if (!jsonNode.path("data").path("command").isMissingNode()) {
+      commandType = CommandType.valueOf(jsonNode.path("data").path("command").asText());
+    }
+
+    if (commandType != CommandType.UPLOAD) {
+      throw new SnowflakeSQLException(
+          new Exception(),
+          SqlState.INTERNAL_ERROR,
+          ErrorCode.INTERNAL_ERROR.getMessageCode(),
+          "This API only supports PUT command");
+    }
+
+    JsonNode locationsNode = jsonNode.path("data").path("src_locations");
+
+    assert locationsNode.isArray();
+
+    final String[] src_locations;
+    final List<RemoteStoreFileEncryptionMaterial> encryptionMaterial;
+    try {
+      src_locations = mapper.readValue(locationsNode.toString(), String[].class);
+      encryptionMaterial = getEncryptionMaterial(commandType, jsonNode);
+    } catch (Exception ex) {
+      throw new SnowflakeSQLException(
+          ex,
+          SqlState.INTERNAL_ERROR,
+          ErrorCode.INTERNAL_ERROR.getMessageCode(),
+          "Failed to parse the locations due to: " + ex.getMessage());
+    }
+
+    final Set<String> sourceFiles;
+    if (commandType == CommandType.UPLOAD) {
+      sourceFiles = expandFileNames(src_locations);
+    } else {
+      sourceFiles = new HashSet<String>(Arrays.asList(src_locations));
+    }
+
+    StageInfo stageInfo = getStageInfo(jsonNode);
+
+    List<SnowflakeFileTransferMetadata> result = new ArrayList<>();
+    if (stageInfo.getStageType() != StageInfo.StageType.GCS
+        && stageInfo.getStageType() != StageInfo.StageType.AZURE
+        && stageInfo.getStageType() != StageInfo.StageType.S3) {
+      throw new SnowflakeSQLException(
+          new Exception(),
+          SqlState.INTERNAL_ERROR,
+          ErrorCode.INTERNAL_ERROR.getMessageCode(),
+          "This API only supports S3/AZURE/GCS");
+    }
+
+    for (String sourceFilePath : sourceFiles) {
+      String sourceFileName = sourceFilePath.substring(sourceFilePath.lastIndexOf("/") + 1);
+      result.add(
+          new SnowflakeFileTransferMetadataV1(
+              stageInfo.getPresignedUrl(),
+              sourceFileName,
+              encryptionMaterial.get(0).getQueryStageMasterKey(),
+              encryptionMaterial.get(0).getQueryId(),
+              encryptionMaterial.get(0).getSmkId(),
```

Comment:
Can we ever have a case where `encryptionMaterial` returns more than one element? If not we may want to assert or at least explain that.

Reply:
encryptionMaterial could have more than one element for a DOWNLOAD operation, but that shouldn't be allowed since we're checking that its UPLOAD in this function.  I don't see asserts being used elsewhere in JDBC so I'll throw a new exception if the size isn't 1

Reply:
`encryptionMaterial` can have more than one element for a `DOWNLOAD`. I'll add an assert